scenes/Global.gd

```diff
@@ -62,18 +73,61 @@ class ScenarioPerformance:
 	
 	# did the player die?
 	var died := false
-	
-	func clear() -> void:
-		seconds = 0
-		lines = 0
-		box_score = 0
-		combo_score = 0
-		score = 0
-		died = false
 
 var scenario := Scenario.new()
 var scenario_performance := ScenarioPerformance.new()
 
+var scenario_history_filename := "user://scenario_history.save"
+var loaded_scenario_history := false
+var scenario_history := {}
+
+# how many records we can store, before we start deleting old ones
+var history_size := 1000
+
+"""
+Records the current scenario performance to the player's history.
+"""
+func add_scenario_history(scenario_name, rank_result) -> void:
+	if scenario_name == "":
+		# can't store history without a scenario name
+		return
+	if !scenario_history.has(scenario_name):
+		scenario_history[scenario_name] = []
+	scenario_history[scenario_name].push_front(rank_result)
+	if scenario_history[scenario_name].size() > history_size:
+		scenario_history[scenario_name].resize(history_size)
+
+func save_scenario_history() -> void:
+	var save_game := File.new()
+	save_game.open(scenario_history_filename, File.WRITE)
+	for key in scenario_history.keys():
+		var rank_results_json = []
+		for rank_result in scenario_history[key]:
+			rank_results_json.append(rank_result.to_dict())
+		save_game.store_line(to_json({
+			"scenario_name" : key,
+			"scenario_history" : rank_results_json
+		}))
+	save_game.close()
+
+func load_scenario_history() -> void:
+	loaded_scenario_history = true
+	
+	var save_game := File.new()
+	if !save_game.file_exists(scenario_history_filename):
+		return
+	
+	save_game.open(scenario_history_filename, File.READ)
+	while !save_game.eof_reached():
+		var current_line = parse_json(save_game.get_line())
+		if current_line == null:
+			continue
+		for rank_result_json in current_line["scenario_history"]:
+			var rank_result = RankCalculator.RankResult.new()
+			rank_result.from_dict(rank_result_json)
+			add_scenario_history(current_line["scenario_name"], rank_result)
+	save_game.close()
+
```

Comment:
Consider splitting this new logic out as 'ScenarioHistory'. We don't want global becoming a blob class.

Reply:
Split logic out as 'ScenarioHistory', which is a singleton. It would be bad if there was more than one in scope.